contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,191 @@
+
+
+    /// Event that is emitted when a token is withdrawn,
+    /// indicating the owner of the collection that it was withdrawn from.
+    ///
+    /// If the collection is not in an account's storage, `from` will be `nil`.
+    ///
+    access(all) event Withdraw(id: UInt64, uuid: UInt64, from: Address?, type: String)
+
+    /// Event that emitted when a token is deposited to a collection.
+    ///
+    /// It indicates the owner of the collection that it was deposited to.
+    ///
+    access(all) event Deposit(id: UInt64, uuid: UInt64, to: Address?, type: String)
+
+    /// Interface that the NFTs must conform to
+    ///
+    access(all) resource interface NFT: ViewResolver.Resolver {
+        /// The unique ID that each NFT has
+        access(all) view fun getID(): UInt64
+
+        // access(all) event ResourceDestroyed(uuid: UInt64 = self.uuid, type: self.getType().identifier)
+
+        /// Get a reference to an NFT that this NFT owns
+        /// Both arguments are optional to allow the NFT to choose
+        /// how it returns sub NFTs depending on what arguments are provided
+        /// For example, if `type` has a value, but `id` doesn't, the NFT 
+        /// can choose which NFT of that type to return if there is a "default"
+        /// If both are `nil`, then NFTs that only store a single NFT can just return
+        /// that. This helps callers who aren't sure what they are looking for 
+        ///
+        /// @param type: The Type of the desired NFT
+        /// @param id: The id of the NFT to borrow
+        ///
+        /// @return A structure representing the requested view.
+        access(all) fun getSubNFT(type: Type, id: UInt64) : &{NonFungibleToken.NFT}? {
+            return nil
+        }
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    access(all) resource interface Provider {
+
+        // We emit withdraw events from the provider interface because conficting withdraw
+        // events aren't as confusing to event listeners as conflicting deposit events
+
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        /// It does not specify whether the ID is UUID or not
+        access(Withdrawable) fun withdraw(withdrawID: UInt64): @{NFT} {
+            post {
+                result.getID() == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+                emit Withdraw(id: result.getID(), uuid: result.uuid, from: self.owner?.address, type: result.getType().identifier)
+            }
+        }
+    }
+
+    /// Interface to mediate deposits to the Collection
+    ///
+    access(all) resource interface Receiver {
+
+        /// deposit takes an NFT as an argument and adds it to the Collection
+        ///
+        access(all) fun deposit(token: @{NFT})
+
+        /// getSupportedNFTTypes returns a list of NFT types that this receiver accepts
+        access(all) view fun getSupportedNFTTypes(): {Type: Bool}
```

Comment:
Personally i do not see much reason for this and the one below it anymore. We have pretty much gone away from a collection storing more then 1 nft type have we not?

Reply:
If anything this should just return a singular type IMHO.

Reply:
This is how it is currently on mainnet. Receivers have no obligation to only accept a single type, even if collections typically only hold one type. If we made this only accept a single type, we wouldn't be able to support use cases like a NFT switchboard and other such uses.